tests/bigquery/pipeline_tests.py

```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# Copyright 2017 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+# in compliance with the License. You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License
+# is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+# or implied. See the License for the specific language governing permissions and limitations under
+# the License.
+
+import google
+import google.auth
+import mock
+import pickle
+import re
+import unittest
+
+
+class TestCases(unittest.TestCase):
+
+  @staticmethod
+  def _create_context():
+    project_id = 'test'
+    creds = mock.Mock(spec=google.auth.credentials.Credentials)
+    return google.datalab.Context(project_id, creds)
+
+  @mock.patch('google.datalab.Context.default')
+  @mock.patch('google.datalab.utils.commands.notebook_environment')
+  @mock.patch('google.datalab.bigquery.Table.exists')
+  @mock.patch('google.datalab.bigquery.commands._bigquery._get_table')
+  def test_pipeline_cell(self, mock_get_table, mock_table_exists, mock_environment,
+                         mock_default_context):
+    table = google.datalab.bigquery.Table('project.test.table')
+    mock_get_table.return_value = table
+    mock_table_exists.return_value = True
+    context = TestCases._create_context()
+    mock_default_context.return_value = context
+
+    env = {
+      'foo_query': google.datalab.bigquery.Query(
+        'SELECT * FROM publicdata.samples.wikipedia LIMIT 5')
+    }
+    mock_environment.return_value = env
+
+    args = {'name': 'bq_pipeline_test', 'debug': True, 'billing': 'foo_billing'}
+    # TODO(rajivpb): The references to foo_query need to be resolved.
+    cell_body = """
+            email: foo@bar.com
+            schedule:
+                start_date: 2009-05-05T22:28:15Z
+                end_date: 2009-05-06T22:28:15Z
+                schedule_interval: '@hourly'
+            input:
+                path: test/path
+                table: project.test.table
+                schema:
```

Comment:
Schema should really be a reference to some file rather than inline IMO. Its unlikely that you'll want to re-specify schema for data in multiple places. It should be a file alongside the data (or somewhere else) you can refer to.

Reply:
https://github.com/googledatalab/pydatalab/issues/529